Foundation/NSLock.swift

```diff
@@ -174,28 +218,47 @@ open class NSRecursiveLock: NSObject, NSLocking {
     
     open func unlock() {
         pthread_mutex_unlock(mutex)
+#if os(OSX) || os(iOS)
+        // Only use _timedLockData if it was initialized by lock(before:)
+        if let tld = _timedLockData {
+            pthread_mutex_lock(_CFGetTimedLockMutex(tld))
+            pthread_cond_broadcast(_CFGetTimedLockCond(tld))
+            pthread_mutex_unlock(_CFGetTimedLockMutex(tld))
+        }
+#endif
     }
     
     open func `try`() -> Bool {
         return pthread_mutex_trylock(mutex) == 0
     }
     
-    open func lock(before limit: Date) {
-        NSUnimplemented()
+    open func lock(before limit: Date) -> Bool {
+        if pthread_mutex_trylock(mutex) == 0 {
+            return true
+        }
+
+        guard var endTime = timeSpecFrom(date: limit) else {
+            return false
```

Comment:
why do we return `false` if `timeSpecFrom(date:)` fails? It should crash here, `false` means the timeout expired.

Reply:
Crashing is a little extreme because a timeout was exceeded. The documentation at https://developer.apple.com/documentation/foundation/nslock/1411133-lock states that the return value is `true if the lock is acquired before limit, otherwise false.`
The `lock(before:)` could have been called with a perfectly valid date in the future but could have expired before it is processed (e.g. if it was something in the near future like 3ns)

Reply:
@spevans sorry, my comment is totally bogus. `timeSpecFrom(date:)` doesn't return `false` if the timeout expired but if it couldn't convert the timeout. That should not be possible. The right fix is obvs to make `timeSpecFrom(date:)` not even return a `Bool` as it should always succeed. Then we also don't need this `return` here.